### ISSUE TITLE
v0.1 Tabular model name refinement discussion

### ISSUE THREAD

**Issue by Innixma** [awslabs/autogluon#782]
For v0.1 release, we are looking to refine and improve the readability of the model names created by Tabular.

Below is a list of proposals. We are looking for community feedback, so please add your thoughts and suggestions!

1. Weighted ensemble renaming:
Current -> weighted_ensemble_k0_l2
Proposed -> WeightedEnsemble_l2
Reasoning: All other models are in CamelCase format except for the weighted ensemble. Additionally, k0 is a remnant from before we were able to handle duplicate model names. (k0 represented the special case = kfold 0, aka train with all the data and use train data as validation data as well).
Alternative 1: 'Weighted_l2'

2. Stacker model renaming:
Current -> {model_name}_STACKER_l1
Proposed -> {model_name}_S_l1
Reasoning: Stacker model names often become very long, such that they are difficult to read and type in general usage. Since we now store information that identifies a model as a stacker elsewhere (such as in .leaderboard or .info), we don't need to be as verbose in the model name itself.
Alternative 1: '_STACK_'
Alternative 2: '_B_' or '_BAG_' due to '_S_' being defined when level != 0 (level > 0 and _S_ is redundant together), and additionally _STACKER_l0 is technically misleading, as the model is not a stacker but actually a bagged ensemble.
Alternative 3: No suffix, since the state of the stack and bag can be found through other means.

3. Classifier / Regressor suffix removal:
Current -> RandomForestClassifier
Proposed -> RandomForest
Reasoning: Classifier / Regressor suffix is a remnant from when we did not have such information available elsewhere. Since it is generally well known by the user what problem type they are dealing with, the extra info added to the model names is generally redundant and complicates the code, especially when handling duplicated model name resolution.

4. Stack level rework:
Current -> {model_name}_l{n}
Proposed -> {model_name}_l{n+1}
Reasoning: level=0 is confusing to users due to an l2 model being the head of a model stack which is 3 layers deep. With this change, an l2 model would be the head of a model stack which is 2 layers deep.

5. FULL rename:
Current -> {model_name}_FULL
Proposed -> {model_name}_F
Reasoning: Same reasoning as Proposal 2 in that it contributes to very long model names.

6. DSTL rename:
Current -> {model_name}_DSTL
Proposed -> {model_name}_D
Reasoning: Same reasoning as Proposal 5.

7. CatBoost rename:
Current -> Catboost
Proposed -> CatBoost
Reasoning: To align with the casing used by the CatBoost project, and to keep consistency with other models casing.

8. KNN rename:
Current: KNeighbors
Proposed -> KNearestNeighbors
Reasoning: Missing Nearest.
Alternative 1: KNNeighbors
Alternative 2: KNN

9. NeuralNet:
Current: NeuralNet
Proposed: ???
Reasoning: NeuralNet might be a strange name to have as we add more and more neural network type models into AG.

**Comment by Innixma** [comment]
My personal recommendations are the following:

1. Accept (Main)
2. Accept Alternative 2 (_BAG_)
3. Accept
4. Accept
5. Reject
6. Reject
7. Accept
8. Accept (Main or Alternative 1)
9. Undecided

With this, the following changes would occur to the example model name:

Current: RandomForestClassifier_STACKER_l0
Proposed: RandomForest_BAG_l1

**Comment by gradientsky** [comment]
1. CamelCase is easier to type, but underscores are easier to read [1], I'd vote for underscores, but with capital L on level since lowecase l is in many cases confused with 1: `weighted_ensemble_L2`.
2. BAG is better because we don't really have stacking in the L0.
3. Accept. Alternative proposal: can we do camel_case for variations of the algorithm? This reads better:
4. Accept, but with capital L to avoid confusion of l with 1 or |.
5. Reject, FULL looks good
6. Reject, DSTL looks good
7. Accept: official project is using CatBoost
8. Accept (main). Alternatively `k_nearest_neighbors` if underscore_case is an option
9. Proposal: tabular_neural_network

Here's example how my options will look like:
```
weighted_ensemble_L2
random_forest_gini_FULL_BAG_L0
random_forest_entr_FULL_BAG_L0
neural_network_FULL_BAG_L0
random_forest_entr_BAG_L0
extra_trees_gini_BAG_L0
extra_trees_entr_BAG_L0
extra_trees_entr_FULL_BAG_L0
extra_trees_gini_FULL_BAG_L0
k_nearest_neighbors_unif_FULL_BAG_L1
k_nearest_neighbors_unif_BAG_L1
k_nearest_neighbors_unif_FULL_BAG_L0
k_nearest_neighbors_unif_BAG_L0
k_nearest_neighbors_dist_BAG_L1
k_nearest_neighbors_dist_BAG_L0
k_nearest_neighbors_dist_FULL_BAG_L1
k_nearest_neighbors_dist_FULL_BAG_L0
```

[1] https://whatheco.de/2013/02/16/camelcase-vs-underscores-revisited/

**Comment by mseeger** [comment]
Default/majority on 1-8

As for 9: Are there not already two tabular neural nets present, one the original one, and the other from FastAI?

**Comment by Innixma** [comment]
@gradientsky Re CamelCase: I feel that CamelCase is better because it is fewer characters and makes the logical distinction between the qualifiers that use underscores (level, _FULL, _BAG, _DSTL, etc.) and the model name itself. This clarifies which part of the name is the model itself and which is the qualities of the model.

Re capital L for level, I am ok with this, the reasoning for lowercase l would be that it generally takes less space than uppercase L, but this is likely not that big of a concern.

**Comment by jwmueller** [comment]
I vote for CamelCase as well and I also vote that level suffix should be capitalized to be more readable (`_L1` instead of `_l1`). 

1. Accept (Main, `WeightedEnsemble_l2`). Another alternative could simply be `Ensemble` since we refer to bagged/stack-ensembles in other ways.

2. I vote to get rid of `_S` and just let the level information `_L1` inform the user when a model is stacker.  To indicate a bagged model, I'm fine if we use either `_BAG` or `_B`, depending on how long our typical model-names end up after all these changes.

3. Accept

4. Accept. I also think docs need to clarify that `_L1` are base models that operate directly on the data features, and feed into stackers.

5. I'm fine with either `_FULL` or `_F`. I guess it should match what suffix we choose for bagged models.

6. I'm fine with either `_DSTL` or `_D`, to match whatever suffix we choose for bagged models.

7. Accept `CatBoost`.

8. Accept `KNearestNeighbors`.  Or if`KNearestNeighbors` looks unnecessarily long with the suffixes, I'm fine to just keep current `KNeighbors`. `KNeighbors` turns up more correct results in Google search than `KNNeighbors`.

9. For our current 2 neural nets, I vote we keep these basic models named as `NeuralNetMxnet`, `NeuralNetTorch`, and use more specific names for more advanced NN architectures that we later add (keeping these names vague seems good to me, since these NN models are not exactly a standard architecture). 

Alternatively, could name them: `TabularNeuralNetMxnet`, `TabularNeuralNetTorch` (but may be too long), or  alternatives to `TabularNeuralNet*` could include: `TabularNN*`, `MLP*`, `MLPNeuralNet*`.

**Comment by jwmueller** [comment]
A major thing is our documentation/tutorial needs to provide a glossary of what all the possible model names & suffixes mean.  Also the `hyperparameters` argument keys should be overloaded to either take the name-abbreviations or the full model names discussed here.

**Comment by Innixma** [comment]
Given we are changing _STACKER suffix, I have another proposal to add:

10. Level suffix in non-bagged models:
Current: CatBoost
Proposed: CatBoost_L1
Reasoning: If we later intend to add a form of multilayer blend stacking to AutoGluon, this will be required to maintain clarity. Additionally, it helps to standardize the level scheme such that any model regardless of if it is bagged will have a level suffix.
Alternative 1: L1 (base) models never have a level suffix.

My personal recommendation is 10. Accept (Main)

**Comment by jwmueller** [comment]
I agree with 10, but this only with stacking activated correct? With no stacking, I don't think there should be `_L1` as it's confusing.

**Comment by Innixma** [comment]
All accepted proposals have been implemented into v0.1 release.
